#중복 데이터
##1. 데이터 불러오기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/1. 데이터 분석을 위한 기초 통계 수학/1. 데이터 클랜징/data/bicycle_dup.csv', engine='python', encoding='cp949')
df

자전거번호              대여일시  대여소번호  ...  이용시간    이용거리   나이
0    SPB-23220   2019-11-01 8:48    646  ...  12.0  1100.0   50
1    SPB-16216   2019-11-04 8:38    646  ...   7.0  1100.0   50
2    SPB-21097   2019-11-04 8:46    646  ...  10.0     NaN   32
3    SPB-22292   2019-11-05 8:34    646  ...  10.0  1380.0   21
4    SPB-07935  2019-11-05 12:29    512  ...  10.0  1650.0  120
..         ...               ...    ...  ...   ...     ...  ...
415  SPB-12116   2019-11-15 4:22    112  ...  11.0  2200.0   22
416  SPB-17768   2019-11-15 9:31    103  ...   5.0     NaN   22
417  SPB-18841  2019-11-15 20:40    113  ...  12.0  2280.0   37
418  SPB-24467  2019-11-15 21:55    152  ...  10.0  1180.0   27
419  SPB-17570  2019-11-15 22:16    391  ...  48.0  8100.0   28

[420 rows x 12 columns]

In [ ]:
print(df[df['이용시간'].isnull()])
print(df[df['이용거리']==2290])
df_group = df.groupby('이용거리')

df_group.mean()

In [ ]:
df_group.mean().loc[42070.0]

대여소번호      116.0
대여거치대        3.0
반납대여소번호    101.0
반납거치대        4.0
이용시간        46.0
나이          22.0
Name: 42070.0, dtype: float64

##2. 데이터 중복 확인
> DataFrame에서는 중복 확인을 위해 duplicated 함수를 지원한다. 중복된 데이터가 나타나면 true로 변환시킨다. 

In [ ]:
df.duplicated(['이용거리'])

>0번과 1번의 이용거리가 같아 1번 이용거리가 True로 설정되었다. 416~418까지도 중복된 데이터이기때문에 True로 설정되었다.

In [ ]:
df.duplicated(['이용거리', '나이'])

>두 가지 이상의 옵션을 설정하게 되면 and로 중복을 확인할 수 있다.

In [ ]:
df.duplicated(['이용거리'], keep='last')

>keep 옵션은 기본값이 first이다. first는 첫번째 값을 기준으로 중복된 데이터가 나오면 true로 변환시키는 것이고 last는 중복된 가장 마지막 값을 false로 지정하는 것이다.

In [ ]:
df.duplicated(['이용거리'], keep=False)
df.duplicated(['이용거리'], keep=False).sum()
# df.duplicated(['이용거리'], keep='last').sum()
# df.duplicated(['이용거리']).sum()

244

keep이 false이면 중복된 데이터를 모두 true로 변환시킨다.

##3. 중복 데이터 제거
>중복된 데이터를 제거하기 위해서는 drop_duplicates() 함수를 이용하면 된다. 사용방법은 중복데이터 확인과 같다.

In [ ]:
df.drop_duplicates(['이용거리'], keep=False)